src/hardware/timer.cpp

```diff
@@ -16,112 +16,221 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include "timer.h"
+
+#include <array>
+#include <cassert>
+#include <cmath>
 
-#include <math.h>
-#include "dosbox.h"
 #include "inout.h"
 #include "pic.h"
 #include "mem.h"
 #include "mixer.h"
-#include "timer.h"
 #include "setup.h"
 #include "support.h"
 
 const std::chrono::steady_clock::time_point system_start_time = std::chrono::steady_clock::now();
 
-static inline void BIN2BCD(uint16_t& val) {
-	const auto b = ((val / 10) % 10) << 4;
-	const auto c = ((val / 100) % 10) << 8;
-	const auto d = ((val / 1000) % 10) << 12;
-	assert(b + c + d <= UINT16_MAX);
+/*
+ Bit 4 and 5    Access mode :
+                0 0 = Latch count value command
+                0 1 = Access mode: lobyte only
+                1 0 = Access mode: hibyte only
+                1 1 = Access mode: lobyte/hibyte
+Ref: https://wiki.osdev.org/Programmable_Interval_Timer
+*/
+enum class AccessMode : uint8_t {
+	Latch = 0b0'0,
+	Low   = 0b0'1,
+	High  = 0b1'0,
+	Both  = 0b1'1
+};
 
-	const uint16_t temp = (val % 10) + static_cast<uint16_t>(b + c + d);
-	val = temp;
+constexpr void decimal_to_bcd(uint16_t &val)
+{
+	const auto first = val % 10;
+	const auto second = (val / 10) % 10;
+	const auto third = (val / 100) % 10;
+	const auto fourth = (val / 1000) % 10;
+	const auto total = first + (second * 16) + (third * 16 * 16) +
+	                   (fourth * 16 * 16 * 16);
+	val = check_cast<uint16_t>(total);
 }
```

Comment:
Can't the stdlib do this for us?

Reply:
The compiler will compute the `16 * 16`s in-place as literals. But if we switched to `pow()`, the compiler can't do that math at compile time, so would be making that library call each time this is evaluated.
(when we move to C++20, then `pow()` and `log()` become available as `constexpr` :+1: )

Reply:
Probably best to leave it as-is then, agreed.